### ISSUE TITLE
UDF support in API

### ISSUE THREAD

**Issue by hannesmuehleisen** [cwida/duckdb#471]
We should support a simple interface to create DuckDB UDFs in the C/C++, Python and R APIs. Now that we have orrified vectors, we can offer two alternatives, one scalar one vectorized. This should also be exposed in the R/Python APIs so we can create UDFs directly from native R/Python functions (like SQLite supports). We may also implement #394 while we are at it.

**Comment by david-jankoski** [comment]
oh this would be amazing to be able to do UDFs from R!

**Comment by tiagokepe** [comment]
Hi folks,

I've just started to think about UDF interfaces to give it a try. I'm following the manual of SQLite3 (https://www.sqlite.org/appfunc.html) to init a draft of the UDF interface in C++.

For simplicity, I've focused only on scalar functions and some result types, then as the interface and the implementation start getting mature we can add more features.

**Create Function Interface**
```
int duckdb_create_function(
  DuckDB *db,
  const char *zFunctionName,
  int nArg,
  void (*xFunc)(ClientContext*, int, duckdb::Value**),
);
```

1. **db** → The 1st parameter is a pointer do the database object to which the SQL function is to be added. 

2. **zFunctionName** → The 2nd parameter is the name of the SQL function in UTF8 (it will be transformed in a string_type, internally). 

3. **nArg** → The 3rd parameter is the number of arguments that the function accepts.
4. **xFunc** → The 4th parameter is a pointer to the callback function.

- Some questions:

1. We will support built-in function overload?

> - SQLLITE: “Built-in functions may be overloaded by new application-defined functions.”
> - SQLITE: “The SQL function creation interfaces may be called multiple times with the same function name. If two calls have the same function number but a different number of arguments, for example, then two variants of the SQL function will be registered, each taking a different number of arguments.”

2. We will support built-in function replacement?
> - I don't think so that this should be allowed.

3. Should UDF functions be inserted into the catalog?
> - SQLITE: “If an application uses more than one database connection then application-defined SQL functions must be added to each database connection separately.”



**Callback Function Interface**
```
static void xFunc(
  ClientContext *context,
  int argc,
  duckdb::Value **argv,
);
```

1. **context** → The 1st parameter, is a pointer to the context from which the SQL function is invoked.
2. **argc** → The 2nd parameter is the number of arguments that this SQL function receives.
3. **argv** → The 3rd parameter is the values for each argument of the SQL function.

The callback function might set its result by invoking one of the following routines:

_duckdb_result_*_ family: **Setting the result of an SQL Function**

    void duckdb_result_blob(ClientContext*, const char*, int n);
    void duckdb_result_varchar(ClientContext*, const char*, int n);
    void duckdb_result_double(ClientContext*, double);
    void duckdb_result_int(ClientContext*, int);
    void duckdb_result_int64(ClientContext*, long long int);
    void duckdb_result_value(ClientContext*, duckdb::Value);
    void duckdb_result_null(ClientContext*);
    void duckdb_result_error(ClientContext*, const char*);

If none of these routines are invoked for a specific callback, then the return value will be NULL. *Other routines for date and timestamp can be created too.

Example of a “noop” scalar function that returns its argument (based on the SQLite3 manual):

```
static void noopfunc(
  ClientContext *context,
  int argc,
  duckdb::Value **argv
){
  assert( argc==1 );
  duckdb_result_value(context, argv[0]);
}
```

**Observation:**
Maybe it's good to couple the _create_function_ within the connection (ClientContext) to be similar with the sqlite3 Python interface, as in this example:

```
def _toTileCase(string):
    return str(string).title()
try:
    sqliteConnection = sqlite3.connect('SQLite_Python.db')
    sqliteConnection.create_function("TOTILECASE", 1, _toTileCase)
```

**Comment by hannesmuehleisen** [comment]
Hey Tiago, I think we would like to also support the SQLite UDF API, but then just implement as-is. See the SQLite API wrapper, the stubs are already there. 

Our 'own' API should allow for vectorized implementation of functions, and it already does that for C++. But we would also like to extend UDF support to Python and R, in which case the parameters have to be converted back and forth between the respective representation and DuckDBs internal format.

**Comment by Mytherin** [comment]
Thanks for the detailed write-up Tiago! Indeed I agree with Hannes, the main user-defined function interface should mirror our current function interface. The primary reason for that is performance. The SQLite style interface would result in a function call invocation for every tuple, which is very slow. Our user-defined function interface should not be slower than our built-in functions.

That means that our user-defined functions should follow the same design interface as the main functions, i.e. operate on a DataChunk as input and take a vector as output:

```cpp
static void udf_function(DataChunk &input, ExpressionState &state, Vector &result);
``` 

We should then be careful that we can support the SQLite UDF API on top of our own C++ API, and perhaps extend our function support so that this can be achieved. Taking a look at the [SQLite create function page](https://www.sqlite.org/c3ref/create_function.html) I think these are the interesting parts:

> The fifth parameter is an arbitrary pointer. The implementation of the function can gain access to this pointer using sqlite3_user_data().

This parameter can be stored inside the ExpressionState of the user-defined function.

> The sixth, seventh and eighth parameters passed to the three "sqlite3_create_function*" functions, xFunc, xStep and xFinal, are pointers to C-language functions that implement the SQL function or aggregate. A scalar SQL function requires an implementation of the xFunc callback only; NULL pointers must be passed as the xStep and xFinal parameters. An aggregate SQL function requires an implementation of xStep and xFinal and NULL pointer must be passed for xFunc. To delete an existing SQL function or aggregate, pass NULL pointers for all three function callbacks.

This seems related to user-defined aggregates, we can allow those as well later, but for now we should focus on scalar functions. In that case, we only need the xFunc which has the format that you describe. We can then implement a wrapper of our C++ UDF API that supports the SQLite function style. This will allow us to support SQLite UDFs transparently, albeit less efficient than UDFs that were made using our own API directly.

> We will support built-in function overload?

This should not be allowed for now, it is not necessary in my opinion and would complicate things.

> We will support built-in function replacement?
> I don't think so that this should be allowed.

I agree, this should not be allowed. 

> Should UDF functions be inserted into the catalog?

As the functions are passed in as a function pointer, we cannot persistently store functions because the function pointers will be invalidated when we exit the program. However, I think having the functions

**Comment by tiagokepe** [comment]
Hi all,

Thanks for our explanation they were very enlightening. I have more thoughts about the C++ UDF API on top of our built-in function API and I wanna share with you (correct me if I'm in the wrong direction or for any misunderstanding).

My idea is from the user perspective and I think it would be nice if the users can create straightforward UDF functions like that:

```
int udf_function(int i) {
	return i;
}
```
For accomplishing that I envision the following API:

```
template<class Ret, class ...Args> void duckdb_create_function(
    string                                    fName,
    vector<duckdb::SQLType>	 arguments,
    duckdb::SQLType                 returnType,
    duckdb::FunctionType          funType,
    Ret                                        (*xFun)(Args...),
    bool                                       useResultVector,
    unique_ptr<FunctionData>   bindInfo
)
```

- **Ret**→ A tampleted parameter that defines the return type of the UDF function pointer (*xFunc);
- **Args**→A Templated parameter(s) that defines the argument types of the function pointer (*xFunc);
- **fName**→The name of the UDF function;
- **arguments**→The SQL type arguments to register the function;
- **returnType**→The SQL return type to register the function;
- **funType**→The type of the UDF function: Unary, Binary or Ternary;

`enum class FunctionType  {UNARY_FUNCTION, BINARY_FUNCTION, TERNARY_FUNCTION};`

- **xFun**→A pointer to the UDF function;
- **useResultVector**→A bool argument to define if we need to pass the result vector as argument to the UDF function;
- **bindInfo**→A pointer to the UDF function;

A prototype of the duckdb_create_function would be:

```
template<class Ret, class ...Args> void duckdb_create_function(
...
) {
	const std::size_t argc = sizeof...(Args);
	switch(argc) {
		case 1:
			auto function = UDFUtils::CreateUnaryScalarFunction<Ret, Args...>(xFun);
			break;
		case 2:
                        auto function = UDFUtils::CreateBinaryScalarFunction<Ret, Args...>(xFun);
			break;
		case 3:
                        auto function = UDFUtils::CreateTernaryScalarFunction<Ret, Args...>(xFun);
			break;
		default:
			break;
	}
	UDFUtils::RegisterFunction(fName, arguments, returnType, function);
}
```

And an implementation of the CreateUnaryScalarFunction is:

```
template<typename Ret, typename Arg, idx_t COUNT=1>
scalar_function_t CreateUnaryScalarFunction(Ret (*func)(Arg))
{
    std::function<Ret(Arg)> f_func = [&](Arg arg0) {
    	return func(arg0);
    };

    std::function<void(DataChunk &, ExpressionState &, Vector &)> scalar_func =
    		[&] (DataChunk &args, ExpressionState &state, Vector &result) {
					UnaryExecutor::Execute<Arg, Ret>(
							args.data[0],
							result,
							COUNT,
							f_func
					);
    };

    return (scalar_function_t &)scalar_func;
}
```

Then, an usage could be (this code compiles):

```
duckdb_create_function<int, int>("udf_function",
  			         {duckdb::SQLType::INTEGER},
			         duckdb::SQLType::INTEGER,
			         FunctionType::UNARY_FUNCTION,
			         &udf_function,
			         false,
			         nullptr
);
```
Maybe I'm very ambitious and some guidelines would be appreciated.

**Comment by talegari** [comment]
Thanks for this great DB and package. R/Python UDFs will be a game-changer for my usecase on a single machine which ideally is

- not large enough for distributed environments (like spark)
- but large enough to hold in RAM
- involve 'groupby' computation with UDFs (so chunked processing is not ideal either!)

Let me know if UDF support is planned to be implemented anytime soon.

**Comment by hannesmuehleisen** [comment]
@talegari what exactly do you mean with groupby? User-defined aggregate functions?

**Comment by talegari** [comment]
@hannesmuehleisen 
By groupby computation, I am looking to compute something like this on a table:

```
duckdb_tbl %>%                      # duckdb_tbl is a tbl object
  group_by(some_variable) %>%
  summarise(x = udf(var_1, var_2))  # lets say udf: (int, int) --> real
```

**Comment by talegari** [comment]
@hannesmuehleisen Will the UDF be useable per-group?

**Comment by polyrand** [comment]
I think having UDFs would be fantastic. I reached this issue while thinking about how to create the simplest possible batch processing system, then DuckDB came to my mind.

This is not directly related to UDFs, but my idea was something like:

Append data to a column "Data.value" -> then execute `UPDATE Data SET result = process(value) WHERE result IS NULL`. Since DuckDB has vectorized / bulk update queries, I think it could be a perfect alternative to other heavier parallel processing frameworks like Dask, Ray, etc. (I'm thinking about Python). Then, multiple python processes could be doing bulk updates without having to do inter-process communication and all that.

**Comment by hawkfish** [comment]
Would it make sense to do something similar for aggregate functions? I'm looking for a way to get my feet wet again...

**Comment by Mytherin** [comment]
We already have support for user-defined aggregates in the C++ API (see e.g. [here](https://github.com/cwida/duckdb/blob/master/test/api/udf_function/test_aggregate_udf.cpp)), but if you want to do something relating to UDFs perhaps a good task would be adding compatibility with SQLite's UDF API to the sqlite3_api_wrapper.

**Comment by hawkfish** [comment]
Got it - I just looked at `connection.hpp` and it didn't look the same so I assumed...  I'll have a look at the `sqlite3_api_wrapper` then.

**Comment by hawkfish** [comment]
Looks like this is the only issue tracking it?

**Comment by Mytherin** [comment]
Yes, but feel free to create a new issue relating to the sqlite3_api_wrapper specifically if you prefer to have a stand-alone one. Then I can assign you there as well.

**Comment by hawkfish** [comment]
This is not straightforward. SQLite UDFs are untyped, so I presume we would need to declare them as `LogicalType::ANY`? Then there is the problem that the function we register in the `Connection` would need to be a shim between `sqlite3_value`s and `DuckDB::Value`s but there is no place to save the user's UDF (e.g., `xFunc`) in the existing UDF APIs.

**Comment by Mytherin** [comment]
That's a good point, I hadn't considered the variable typing. That indeed complicates the functions somewhat, particularly on the output side. On the input side this is not as difficult since we can take whatever vectors are passed in and translate the individual tuples to `sqlite3_value`s. On the output side, we likely just need to output strings since we can't assume the resulting values will have any particular type.

As for the user data, indeed, the existing UDF APIs should be extended to allow for storing user data alongside the function.

**Comment by hannesmuehleisen** [comment]
We now have support for Python map-style UDFs: #1569

**Comment by hawkfish** [comment]
I've been looking at implementing `array_slice` but I'm not clear on how to share the auxiliary data between input and columns.

**Comment by ParadoxShmaradox** [comment]
Can UDF be extended to support the Java API as well? I assume it won't be vectorized and that's fine.
My use case is pretty simple, adding a CONVERT_TZ like function (from mysql) to translate timestamps to something human readable both in a where clause and in the select.

**Comment by hawkfish** [comment]
> Can UDF be extended to support the Java API as well? I assume it won't be vectorized and that's fine.
> My use case is pretty simple, adding a CONVERT_TZ like function (from mysql) to translate timestamps to something human readable both in a where clause and in the select.

By "human readable" do you mean "in a time zone the user understands"? Or do you mean "using a date order from the user's locale"?

**Comment by ParadoxShmaradox** [comment]
"in a time zone the user understands".
Lets say the data is saved in UTC timestamps but the context of the query is to select one day so you would write:

SELECT *, CONVERT_TZ(`TIMESTAMP`, 'UTC' , 'America/Chicago') AS date FROM some_table WHERE 
id = 12345 and
`TIMESTAMP` >= CONVERT_TZ('2017-09-29 00:00:00' , 'America/Chicago', 'UTC') AND  `TIMESTAMP` < CONVERT_TZ('2017-09-30 00:00:00' , 'America/Chicago', 'UTC');

This way it's "easy" to just change the date to query on and view the returned data without having to deal with date to timestamp conversion outside the query execution context.

I'm assuming that duckdb will never handle time zones (although it would be great) because it adds too much complexity and size to the artifact and that should be the user responsibility (for example to update your IANA table).

**Comment by hawkfish** [comment]
Thanks for the example/clarification. I'm actually asking here because temporal analytics is one of my things and I'd like to understand use cases. 

The one you describe is a common one where the data is stored in UTC (as it should be) but the user needs the data to be filtered for a particular locale for reporting (e.g., financials must legally be reported in a particular time zone) or translated for "anchoring" (e.g., the user is a sysop who got woken up at 0315 because a server went down at 0257 their time and they want to look at the log messages without having to do mental math). `CONVERT_TZ` allows these translations to be pushed down to the DB.

One problem with this approach is that it is not clear what the function even does. A timestamp is just a point on a physical axis measured in UTC, so the operations of converting to and from UTC are completely different beasts:

* Converting *to* UTC takes a presentation layer representation (i.e., a string) and converts it to a timestamp. 
* Converting *from* UTC takes a timestamp and produces a presentation layer string. 

`CONVERT_TZ` is thus an impossible function(!) but the way MySQL implements it is to do something strange with their unique internal representation (which is a place-value system that even allows months and days to be zero) and then hope that the way it comes out at the other end still makes sense when displayed.

All this is a bit pedantic, but the bottom line is I'd strongly recommend keeping presentation layer operations out of the database. Your Java function is great where it is and I'd keep it that way. It will also make your code more portable.

That said, I think that there might be a place for two functions that parse and format temporal string representations with TZ support. That still requires some large, ugly tables (time zone boundaries and rules change surprisingly often) but a UDF module with the tables embedded (or read from disk) could be very useful. In a previous system I worked on, we used [ICU](https://unicode-org.github.io/icu/userguide/datetime/) for this and maybe there is a way to build a UDF from the ICU library.

**Comment by hawkfish** [comment]
> I'm assuming that duckdb will never handle time zones (although it would be great) because it adds too much complexity and size to the artifact and that should be the user responsibility (for example to update your IANA table).

Not only that, but the analytic use cases for time zones are not that common (which is why most databases don't support them). These use cases tend to be things that require extracting date parts for the calendar/locale the data originated from (e.g., find all sales that happened on Monday mornings anywhere in the world) which involves performing a calendar computation on each row. That is so slow one might want to partition the data and share the calendar between similar rows(!)

**Comment by ParadoxShmaradox** [comment]
Yes, you could always use 2 UDF functions one to convert from a date representation and one to convert to, that's totally fine.

I'm building a JDBC Driver that's under the hood talks to some service that will use duckdb. In this case my presentation layer is some SQL IDE like dbeaver, so my logic is either in the driver (which I want to keep dumb) or that service.

I can implement UDFs myself but then I'd have to cut that function from the select (but add the column the function needs) send that query to duckdb and then run the function on the returned data preserving the original sql columns. I already do something like that with the where clause to convert it initially to a native timestamp value using JSqlParser as an sql parser.

I agree that time zone based analytics on large data cross timezone datasets is a corner case and there are ways to mitigate them like splitting into multiple queries or pre computing fields but for my simple use case I think it's fine but again as long as there is UDF support in Java I can happily shoot myself in the leg slightly to make it work :)

**Comment by bigfacewo** [comment]
Can you consider supporting UDF for Java? I used H2 database before and defined many UDFs. If duckdb can support Java UDF, it will reduce the work of rewriting UDF.

**Comment by hannes** [comment]
We have UDFs now, they are supported in C++, Python (somewhat), Node.js/WASM and soon R. Closing this issue for more detailed issues regarding individual integrations.

**Comment by mgttt** [comment]
FYI, using pandas dataframe seems leading to somehow memleak...

**Comment by Alex-Monahan** [comment]
Could you please open a new issue and provide a reproducible example? It helps us fix the issue and add a test to prevent it reoccurring!

**Comment by jeremyhermann** [comment]
> We have UDFs now, they are supported in C++, Python (somewhat), Node.js/WASM and soon R. Closing this issue for more detailed issues regarding individual integrations.

Are Python UDFs supported now? I can't find them in the docs or source code

**Comment by Mause** [comment]
> > We have UDFs now, they are supported in C++, Python (somewhat), Node.js/WASM and soon R. Closing this issue for more detailed issues regarding individual integrations.
> 
> @hannes, are Python UDFs supported now? I can't find them in the docs or source code

They are (somewhat) supported via the [`map`](https://duckdb.org/docs/api/python/reference/#duckdb.DuckDBPyRelation.map) method

**Comment by buremba** [comment]
@Mause is there any code example? I see the method but I can't make it work. I'm pretty sure that I'm not using it right but I wonder if it's possible to map Python functions as UDF or UDTFs?

```
import duckdb

# connect to an in-memory database
con = duckdb.connect()
def test(*args, **kwargs):
    return []

con.table("test").map(test)

con.query("select count(*) from test")
```

Error:
```
duckdb.Error: Table does not exist!
```

**Comment by mnjdhl** [comment]
Do we have UDF support for Rust?

**Comment by szarnyasg** [comment]
@mnjdhl not at the moment, see https://github.com/duckdb/duckdb-rs/issues/181#issuecomment-1636841886

**Comment by ran-codes** [comment]
Are R UDF's on the roadmap?

**Comment by DerZc** [comment]
Do you have UDF support for Java?

**Comment by uwemaurer** [comment]
> Do you have UDF support for Java?

No, but planned. https://github.com/duckdb/duckdb-java/issues/256
